gremlin-go/driver/connectionPool.go

```diff
@@ -0,0 +1,168 @@
+/*
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+*/
+
+package gremlingo
+
+import (
+	"crypto/tls"
+	"sync"
+)
+
+type connectionPool interface {
+	write(*request) (ResultSet, error)
+	close()
+}
+
+type loadBalancingPool struct {
+	url        string
+	authInfo   *AuthInfo
+	tlsConfig  *tls.Config
+	logHandler *logHandler
+
+	newConnectionThreshold int
+	connections            []*connection
+	loadBalanceLock        sync.Mutex
+}
+
+func (pool *loadBalancingPool) close() {
+	for _, connection := range pool.connections {
+		err := connection.close()
+		if err != nil {
+			pool.logHandler.logf(Warning, errorClosingConnection, err.Error())
+		}
+	}
+}
+
+func (pool *loadBalancingPool) write(request *request) (ResultSet, error) {
+	connection, err := pool.getLeastUsedConnection()
+	if err != nil {
+		return nil, err
+	}
+	return connection.write(request)
+}
+
+func (pool *loadBalancingPool) getLeastUsedConnection() (*connection, error) {
+	pool.loadBalanceLock.Lock()
+	defer pool.loadBalanceLock.Unlock()
+	if len(pool.connections) == 0 {
+		return pool.newConnection()
+	} else {
+		var leastUsed *connection = nil
+		validIndex := 0
+		for _, connection := range pool.connections {
+			// Purge dead connections from pool
+			if connection.state == established {
+				// Close and purge connections from pool if there is more than one being unused
+				if leastUsed != nil && (leastUsed.activeResults() == 0 && connection.activeResults() == 0) {
+					// Close the connection asynchronously since it is a high-latency method
+					go func() {
+						pool.logHandler.log(Info, closeUnusedPoolConnection)
+						err := connection.close()
```

Comment:
No, this is not possible because the closing connection has already been purged from the pool by the time goroutine2 has accessed `getLeastUsedConnection`